### PR TITLE
Feature sdk 297 v4 auth

### PR DESCRIPTION
**Pull request by victoriaxu098**
There are two methods to express v4 authentication information:
1. HTTP Authorization header. 
2. Query string parameters(presigning a URL). 

Signing steps:
1. Create a canonical request for Signature Version 4
Arrange the contents of your request (host, action, headers, etc.) into a standard (canonical) format. The canonical request is one of the inputs used to create a string to sign.
2. Create a string to sign for Signature Version 4
Create a string to sign with the canonical request and extra information such as the algorithm, request date, credential scope, and the digest (hash) of the canonical request.
3. Calculate the signature for AWS Signature Version 4
Derive a signing key by performing a succession of keyed hash operations (HMAC operations) on the request date, Region, and service, with your AWS secret access key as the key for the initial hashing operation. After you derive the signing key, you then calculate the signature by performing a keyed hash operation on the string to sign. Use the derived signing key as the hash key for this operation.
4. Add the signature to the HTTP request

### CODE REVIEW

File: src/main/java/com/emc/object/s3/S3SignerV4.java [hunk 1]
```diff
@@ -0,0 +1,392 @@
...
+                if (parameter.getValue() != null)
+                    sortedQueryString.append(parameter.getValue());
+            }
+            queryString.append(sortedQueryString).append("\n");
+        } else
+            queryString.append("\n");
+        return queryString.toString();
+    }
+
+    protected SortedMap<String, String> getCanonicalizedHeaders(Map<String, List<Object>> headers,
```
Comment (xiaoxin-ren) [inline src/main/java/com/emc/object/s3/S3SignerV4.java:190]: According to the comment, this function will add x-emc-/x-amz- headers but real function is to exclude excludedSignedHeaders, what's the real desire? Is it possible to consolidate the function into the one in S3Signer.java?

File: src/main/java/com/emc/object/s3/S3SignerV4.java [hunk 2]
```diff
@@ -0,0 +1,392 @@
...
+            date = getDate(parameters, headers);
+            request.getHeaders().add(S3Constants.AMZ_DATE, date);
+            request.getHeaders().add(RestUtil.HEADER_HOST, hostHeader);
+        }
+
+        String shortDate = getShortDate(date);
+
+        RestUtil.putSingle(headers, S3Constants.AMZ_DATE, date);
+        RestUtil.putSingle(headers, RestUtil.HEADER_HOST, hostHeader);
+        headers = request.getHeaders();
```
Comment (xiaoxin-ren) [inline src/main/java/com/emc/object/s3/S3SignerV4.java:57]: Is it possible to extract them into a function, e.g. processHeaders() to include all header changes in v4 signer? e.g. Region information as a lowercase alphanumeric string, etc
https://docs.aws.amazon.com/general/latest/gr/sigv4_changes.html

File: src/main/java/com/emc/object/s3/S3SignerV4.java [hunk 3]
```diff
@@ -0,0 +1,392 @@
...
+        byte[] hash = hash256(canonicalRequest);
+        String hashedRequest = hexEncode(hash);
+        SortedMap<String, String> canonicalizedHeaders = getCanonicalizedHeaders(headers, parameters);
+        StringBuilder signedHeaders = new StringBuilder();
+        for (String name : canonicalizedHeaders.keySet()) {
+            if (signedHeaders.length() != 0)
+                signedHeaders.append(";");
+            signedHeaders.append(name);
+        }
+        String stringToSign = getStringToSign(request.getMethod(), resource, parameters, headers, date, serviceType) + hashedRequest;
```
Comment (xiaoxin-ren) [inline src/main/java/com/emc/object/s3/S3SignerV4.java:73]: A decent abstract would not require appending "hashedRequest" but a complete sign string returned by the function. Is it possible to leverage s3Config.isUseV2Signer() and put everything inside the method in parent class?

**Review by xiaoxin-ren — COMMENTED** [review]
Generally look good to me, left a few questions regarding class abstract between V2 and V4 Signer.

File: src/main/java/com/emc/object/s3/S3SignerV4.java (hunk 1 above)
Comment (victoriaxu098) [reply src/main/java/com/emc/object/s3/S3SignerV4.java:188]: Removed unused comments. This method is already an overwrite method. V4 has different behavior on this with v2.

File: src/main/java/com/emc/object/s3/S3SignerV4.java (hunk 2 above)
Comment (victoriaxu098) [reply src/main/java/com/emc/object/s3/S3SignerV4.java]: Thanks for suggestion. Have put adding header logic into method addHeadersForV4( ).

File: src/main/java/com/emc/object/s3/S3SignerV4.java (hunk 3 above)
Comment (victoriaxu098) [reply src/main/java/com/emc/object/s3/S3SignerV4.java]: hashedRequest is requied in v4 stringToSign. Have integrated hashedRequest into stringToSign for v4 signer.

File: src/main/java/com/emc/object/s3/jersey/S3JerseyClient.java [hunk 4]
```diff
@@ -200,14 +202,27 @@ public S3JerseyClient(S3Config config, ClientHandler clientHandler) {
             }
         }
 
+        ClientHandler handler = client.getHeadHandler();
```
Comment (twincitiesguy) [inline src/main/java/com/emc/object/s3/jersey/S3JerseyClient.java:205]: need a comment here explaining that we have to re-order the filters and why it is necessary

File: src/main/java/com/emc/object/util/RestUtil.java [hunk 5]
```diff
@@ -206,6 +211,23 @@ public static Date headerParse(String dateString) {
         }
     }
 
+    public static Date headerParse(String dateString, String headerKey) {
+        if (dateString == null) return null;
+        try {
+            if(headerKey.equals(S3Constants.AMZ_DATE)) {
+                // convert date
+                SimpleDateFormat sdf = new SimpleDateFormat(AMZ_DATE_FORMAT);
```
Comment (twincitiesguy) [inline src/main/java/com/emc/object/util/RestUtil.java:219]: this is an expensive operation - we should use a `ThreadLocal` reference, like the other dateformats at the top of this class - in fact, it may be better to use a static initializer, or initialize in `ThreadLocal.withInitial()`, and get rid of the `getHeaderFormat()` static method

File: src/main/java/com/emc/object/util/RestUtil.java [hunk 6]
```diff
@@ -206,6 +211,23 @@ public static Date headerParse(String dateString) {
         }
     }
 
+    public static Date headerParse(String dateString, String headerKey) {
```
Comment (twincitiesguy) [inline src/main/java/com/emc/object/util/RestUtil.java:214]: I don't see a reason to have the `headerKey` param - we have `headerParse()` method for non-AMZ parsing - this method can be called `amzHeaderParse()` for AMZ-specific header parsing

File: src/main/java/com/emc/object/s3/S3SignerV2.java [hunk 7]
```diff
@@ -26,205 +26,54 @@
...
+        super(s3Config);
+    }
+
+    public void sign(ClientRequest request, String resource, Map<String, String> parameters, Map<String, List<Object>> headers) {
+        String stringToSign = getStringToSign(request.getMethod(), resource, parameters, headers);
+        String signature = getSignature(stringToSign, null);
+        RestUtil.putSingle(headers, "Authorization", "AWS " + s3Config.getIdentity() + ":" + signature);
     }
 
+    // this method is for testing
```
Comment (twincitiesguy) [inline src/main/java/com/emc/object/s3/S3SignerV2.java:48]: what calls this method?  Can we refactor the test so it uses another method, so we don't have methods that only exists for testing?  seems like the test could use the other `sign()` method and construct an appropriate `ClientRequest` instance

File: src/main/java/com/emc/object/s3/S3SignerV4.java [hunk 8]
```diff
@@ -0,0 +1,397 @@
...
+        // #3 Calculate the signature for AWS Signature Version 4
+        byte[] key = getSigningKey(shortDate, serviceType);
+        String signature = getSignature(stringToSign, key);
+        log.debug("Signature: {}", signature);
+
+        // #4 Adding signing information to the authorization header
+        RestUtil.putSingle(headers, "Authorization", S3Constants.AWS_HMAC_SHA256_ALGORITHM +
+                " Credential=" + s3Config.getIdentity() + "/" + shortDate +
+                "/" + S3Constants.AWS_DEFAULT_REGION + "/" + serviceType + "/" + S3Constants.AWS_V4_TERMINATOR +
+                ", SignedHeaders=" + signedHeaders.toString() + ", " + S3Constants.PARAM_SIGNATURE + "= " + signature);
```
Comment (twincitiesguy) [inline src/main/java/com/emc/object/s3/S3SignerV4.java:71]: there is a space after the equals here - is that going to cause any problems?

File: src/main/java/com/emc/object/s3/S3SignerV4.java [hunk 9]
```diff
@@ -0,0 +1,397 @@
...
+            signedHeaders.append(name);
+        }
+        canonicalRequest.append("\n");
+
+        signedHeaders.append("\n");
+        canonicalRequest.append(signedHeaders);
+
+        String hashedPayload = "";
+        String payload = "";
+        byte[] hash = hash256(payload);
```
Comment (twincitiesguy) [inline src/main/java/com/emc/object/s3/S3SignerV4.java:159]: this will always be the same hash (of an empty string) - why is this necessary?

also, can we combine the two `getCanonicalRequest()` methods and add a boolean param whether to include the hash?  30 lines of duplicate code...

File: src/main/java/com/emc/object/s3/S3SignerV4.java [hunk 10]
```diff
@@ -0,0 +1,397 @@
...
+        }
+        // use Date header
+        date = RestUtil.getFirstAsString(headers, RestUtil.HEADER_DATE);
+        if (date == null) {
+            // must have a date in the headers
+            date = RestUtil.getRequestDate(s3Config.getServerClockSkew());
+        }
+
+        // convert date
+        SimpleDateFormat sdf = new SimpleDateFormat(HEADER_DATE_FORMAT, Locale.US);
```
Comment (twincitiesguy) [inline src/main/java/com/emc/object/s3/S3SignerV4.java:251]: avoid constructing dateformats for every request - this will be a perf bottleneck - use `ThreadLocal` if you must reformat any `Date`s

File: src/main/java/com/emc/object/s3/S3SignerV4.java [hunk 11]
```diff
@@ -0,0 +1,397 @@
...
+        } catch (ParseException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    protected String getShortDate(String date) {
+        // Date must be consistent with timestamp, so extract it
+        // from previous date time format instead of get current date
+        String shortDate = "";
+        SimpleDateFormat sdf = new SimpleDateFormat(AMZ_DATE_FORMAT, Locale.US);
```
Comment (twincitiesguy) [inline src/main/java/com/emc/object/s3/S3SignerV4.java:265]: another SimpleDateFormat constructor - use `ThreadLocal`s or switch to Java 8 `DateTime`

File: src/main/java/com/emc/object/s3/S3SignerV4.java [hunk 12]
```diff
@@ -0,0 +1,397 @@
...
+            parameters.put(override.getQueryParam(), headerOverrides.get(override));
+        }
+
+        Map<String, List<Object>> headers = request.getHeaders();
+        String method = request.getMethod().name();
+
+        String date = null;
+        String serviceType = getServiceType();
+        if (headers.containsKey(S3Constants.AMZ_DATE)) {
+            date = RestUtil.getFirstAsString(headers, S3Constants.AMZ_DATE);
```
Comment (twincitiesguy) [inline src/main/java/com/emc/object/s3/S3SignerV4.java:327]: `getDate()` already checks for `AMZ_DATE`, so why are we checking again here?

File: src/main/java/com/emc/object/s3/jersey/ErrorFilter.java [hunk 13]
```diff
@@ -55,10 +55,10 @@ public ClientResponse handle(ClientRequest request) throws ClientHandlerExceptio
 
             // check for clock skew (can save hours of troubleshooting)
             if (response.getStatus() == 403) {
-                Date clientTime = RestUtil.headerParse(RestUtil.getFirstAsString(request.getHeaders(), S3Constants.AMZ_DATE));
+                Date clientTime = RestUtil.headerParse(RestUtil.getFirstAsString(request.getHeaders(), S3Constants.AMZ_DATE), S3Constants.AMZ_DATE);
                 if (clientTime == null)
-                    clientTime = RestUtil.headerParse(RestUtil.getFirstAsString(request.getHeaders(), RestUtil.HEADER_DATE));
-                Date serverTime = RestUtil.headerParse(RestUtil.getFirstAsString(response.getHeaders(), RestUtil.HEADER_DATE));
+                    clientTime = RestUtil.headerParse(RestUtil.getFirstAsString(request.getHeaders(), RestUtil.HEADER_DATE), RestUtil.HEADER_DATE);
```
Comment (twincitiesguy) [inline src/main/java/com/emc/object/s3/jersey/ErrorFilter.java:60]: seems like `HEADER_DATE` should **not** parse with `AMZ` format

File: src/main/java/com/emc/object/s3/S3Signer.java [hunk 14]
```diff
@@ -0,0 +1,219 @@
...
+        if (s3Config.isSignMetadataSearch()) {
+            signedParameters.add(S3Constants.PARAM_QUERY);
+            signedParameters.add(S3Constants.PARAM_SEARCH_METADATA);
+        }
+    }
+
+    public abstract void sign(ClientRequest request, String resource, Map<String, String> parameters,
+                              Map<String, List<Object>> headers);
+
+    protected abstract String getSignature(String stringToSign, byte[] signingKey);
```
Comment (twincitiesguy) [inline src/main/java/com/emc/object/s3/S3Signer.java:52]: need docs on these abstract methods - what is the method contract?  i.e. what happens if you pass in `null` to `signingKey`?

**Review by twincitiesguy — CHANGES_REQUESTED** [review]

File: src/main/java/com/emc/object/s3/S3Config.java [hunk 15]
```diff
@@ -258,6 +260,19 @@ public void setSignMetadataSearch(boolean signMetadataSearch) {
...
+    @ConfigUriProperty
+    public boolean isUseV2Signer() {
+        return useV2Signer;
+    }
+
+    /**
+     * Sets the AWS signature version for authentication requests, default is v4,
+     * set this to true to use v2.
+     */
+    public void setUseV2Signer(boolean useV2Signer) {
```
[the remark this answers is not in the document]
Comment (victoriaxu098) [reply src/main/java/com/emc/object/s3/S3Config.java:280]: Added withUseV2Signer().

File: src/main/java/com/emc/object/s3/S3SignerV2.java [hunk 16]
```diff
@@ -26,205 +26,54 @@
...
-                S3Constants.PARAM_ENDPOINT,
-                S3Constants.PARAM_IS_STALE_ALLOWED));
-        if (s3Config.isSignMetadataSearch()) {
-            signedParameters.add(S3Constants.PARAM_QUERY);
-            signedParameters.add(S3Constants.PARAM_SEARCH_METADATA);
-        }
+        super(s3Config);
+    }
+
+    public void sign(ClientRequest request, String resource, Map<String, String> parameters, Map<String, List<Object>> headers) {
```
[the remark this answers is not in the document]
Comment (victoriaxu098) [reply src/main/java/com/emc/object/s3/S3SignerV2.java:66]: Added.

File: src/main/java/com/emc/object/s3/S3Config.java [hunk 17]
```diff
@@ -82,6 +82,7 @@ protected static int defaultPort(Protocol protocol) {
     protected int retryBufferSize = DEFAULT_RETRY_BUFFER_SIZE;
     protected float faultInjectionRate = 0.0f;
     protected boolean signMetadataSearch = true;
+    protected boolean useV2Signer = false;
```
[the remark this answers is not in the document]
Comment (victoriaxu098) [reply src/main/java/com/emc/object/s3/S3Config.java]: Have switched default signer to V2.

File: src/test/resources/test.properties.template [hunk 18]
```diff
@@ -17,3 +17,6 @@ s3.endpoint=http[s]://<ecshost>[:9020|:9021]
 
 ## Optional -- HTTP proxy.  Useful for debugging.
 #http.proxyUri=http://localhost:8888
+
+## Whether to use v2 signer, uncomment to use v2 signer instead of v4 signer
+#useV2Signer=false
```
[the remark this answers is not in the document]
Comment (victoriaxu098) [reply src/test/resources/test.properties.template]: Removed the parameter and added V4 tests for all client unit tests.

File: src/test/java/com/emc/object/s3/S3JerseyClientTest.java [hunk 19]
```diff
@@ -2224,6 +2242,45 @@ public void testPreSignedPutUrl() throws Exception {
                         "?AWSAccessKeyId=AKIAIOSFODNN7EXAMPLE&Expires=1175139620&Signature=kPVlidlScN00QlwJMeLd9YWmpOw%3D",
                 url.toString());
 
+        s3Config = super.createS3Config();
```
[the remark this answers is not in the document]
Comment (victoriaxu098) [reply src/test/java/com/emc/object/s3/S3JerseyClientTest.java:2623]: Added tests for v4 in S3JerseyClientV4Test.java.

File: src/test/java/com/emc/object/s3/S3JerseyClientTest.java [hunk 20]
```diff
@@ -2224,6 +2242,45 @@ public void testPreSignedPutUrl() throws Exception {
...
+                    .put(content);
+            Assert.assertEquals(content, client.readObject(getTestBucket(), key, String.class));
+            S3ObjectMetadata metadata = client.getObjectMetadata(getTestBucket(), key);
+            Assert.assertEquals("bar", metadata.getUserMetadata("foo"));
+        }
+    }
+
+    @Test
+    public void testPreSignedPutUrlV4() throws Exception {
+        S3Config s3Config = new S3Config(new URI("https://s3.amazonaws.com")).withUseVHost(true)
```
[the remark this answers is not in the document]
Comment (victoriaxu098) [reply src/test/java/com/emc/object/s3/S3JerseyClientTest.java]: Have made the tests seperated for v2 and v4.

File: src/test/java/com/emc/object/s3/AbstractS3ClientTest.java [hunk 21]
```diff
@@ -103,6 +103,7 @@ protected static S3Config s3ConfigFromProperties() throws Exception {
         URI endpoint = new URI(TestConfig.getPropertyNotEmpty(props, TestProperties.S3_ENDPOINT));
         boolean enableVhost = Boolean.parseBoolean(props.getProperty(TestProperties.ENABLE_VHOST));
         String proxyUri = props.getProperty(TestProperties.PROXY_URI);
+        boolean useV2Signer = Boolean.parseBoolean(props.getProperty(TestProperties.USE_V2_SIGNER));
```
[the remark this answers is not in the document]
Comment (victoriaxu098) [reply src/test/java/com/emc/object/s3/AbstractS3ClientTest.java]: Removed.

File: src/main/java/com/emc/object/util/RestUtil.java [hunk 22]
```diff
@@ -179,6 +183,7 @@ public static String generateRawQueryString(Map<String, String> parameterMap) {
             while (paramI.hasNext()) {
                 String name = paramI.next();
                 query.append(urlEncode(name));
+                // this may be an issue in v4... amazon wants an empty string if there is no value
```
[the remark this answers is not in the document]
Comment (victoriaxu098) [reply src/main/java/com/emc/object/util/RestUtil.java]: Verified and removed the comments. '= ' and empty string are only needed when calculating the signature in v4.

File: src/main/java/com/emc/object/s3/request/PresignedUrlRequest.java [hunk 23]
```diff
@@ -68,6 +70,10 @@ public Date getExpirationTime() {
         return expirationTime;
     }
 
+    public Method getMethod() {
```
[the remark this answers is not in the document]
Comment (victoriaxu098) [reply src/main/java/com/emc/object/s3/request/PresignedUrlRequest.java:73]: Added.

File: src/main/java/com/emc/object/s3/jersey/S3JerseyClient.java [hunk 24]
```diff
@@ -141,7 +140,10 @@ public S3JerseyClient(S3Config s3Config) {
     public S3JerseyClient(S3Config config, ClientHandler clientHandler) {
         super(new S3Config(config)); // deep-copy config so that two clients don't share the same host lists (SDK-122)
         s3Config = (S3Config) super.getObjectConfig();
-        this.signer = new S3SignerV2(s3Config);
+        if(!s3Config.isUseV2Signer())
```
[the remark this answers is not in the document]
Comment (victoriaxu098) [reply src/main/java/com/emc/object/s3/jersey/S3JerseyClient.java]: Changed.

File: src/main/java/com/emc/object/s3/jersey/S3JerseyClient.java (hunk 4 above)
Comment (victoriaxu098) [reply src/main/java/com/emc/object/s3/jersey/S3JerseyClient.java:205]: Added comments.

File: src/main/java/com/emc/object/util/RestUtil.java (hunk 6 above)
Comment (victoriaxu098) [reply src/main/java/com/emc/object/util/RestUtil.java]: Changed.

File: src/main/java/com/emc/object/s3/S3SignerV2.java (hunk 7 above)
Comment (victoriaxu098) [reply src/main/java/com/emc/object/s3/S3SignerV2.java]: Changed.

File: src/main/java/com/emc/object/s3/S3SignerV4.java (hunk 8 above)
Comment (victoriaxu098) [reply src/main/java/com/emc/object/s3/S3SignerV4.java]: Removed the space.

File: src/main/java/com/emc/object/s3/S3SignerV4.java (hunk 9 above)
Comment (victoriaxu098) [reply src/main/java/com/emc/object/s3/S3SignerV4.java]: 1. Made changes to initializeaa static final value for hashed payload. 
2. Combined the two getCanonicalRequest() methods.

File: src/main/java/com/emc/object/s3/S3SignerV4.java (hunk 12 above)
Comment (victoriaxu098) [reply src/main/java/com/emc/object/s3/S3SignerV4.java:317]: Removed the first check.

File: src/main/java/com/emc/object/s3/jersey/ErrorFilter.java (hunk 13 above)
Comment (victoriaxu098) [reply src/main/java/com/emc/object/s3/jersey/ErrorFilter.java]: Refactored RestUtil.headerParse() method, removed the second parameter.

File: src/main/java/com/emc/object/s3/S3Signer.java (hunk 14 above)
Comment (victoriaxu098) [reply src/main/java/com/emc/object/s3/S3Signer.java:42]: Added java docs for these methods.

File: src/main/java/com/emc/object/s3/S3SignerV4.java (hunk 10 above)
Comment (victoriaxu098) [reply src/main/java/com/emc/object/s3/S3SignerV4.java]: Used DateTimeFormatter to replace SimpleDateFormat

File: src/main/java/com/emc/object/s3/S3SignerV4.java (hunk 11 above)
Comment (victoriaxu098) [reply src/main/java/com/emc/object/s3/S3SignerV4.java]: Used DateTimeFormatter to replace SimpleDateFormat

File: src/main/java/com/emc/object/util/RestUtil.java (hunk 5 above)
Comment (victoriaxu098) [reply src/main/java/com/emc/object/util/RestUtil.java]: Changed.

**Review by dongyeh — APPROVED** [review]

**Review by twincitiesguy — APPROVED** [review]
